brave/src/main/java/brave/RealSpan.java

```diff
@@ -139,17 +142,30 @@ final class RealSpan extends Span {
   }
 
   @Override public void finish(long timestamp) {
-    synchronized (state) {
+    lock.lock();
```

Comment:
@codefromthecrypt please correct me if I am doing something silly here, but it looks like operations over `pendingSpans` do not depend on `state` and could be synchronized over a separate `lock`.

Reply:
state is a MutableSpan looked up by tracecontext. The problem with us having to remove all synchronized is now we have a much more complex sitation. I mean I am assuming it is really very required and still have a hard time believing it is.

Before the object was its own lock, guarded by itself. Now we are externalizing it even though the object (MutableSpan) can be mutated via various callers.

So, I think we need to focus on what is the pattern suggested for migrating to ReentrantLock from synchronized, when in synchronized we can use the object itself as the lock. Be careful to not expose API surface when thinking this through!

One way is to make sure that any locked access is using PendingSpan (internal) and moving the lock there. At least there, it has 1-1 relationship to what it is guarding (a mutable span)

Reply:
> One way is to make sure that any locked access is using PendingSpan (internal) and moving the lock there. At least there, it has 1-1 relationship to what it is guarding (a mutable span)

Got it, thanks @codefromthecrypt . To elaborate a bit on the issue, `synchronized` over `state` is fine and non-issue with virtual thread (it is super short and I didn't change it expect here) but `pendingSpans.xxx` methods have unpredictable latency and may hold `synchronized`  for a looong time - this is an issue. 

> Before the object was its own lock, guarded by itself. Now we are externalizing it even though the object (MutableSpan) can be mutated via various callers.

That's I think is one of the problems: we cannot reliably guard `MutableSpan` unless all accesses to it are guarded as well (not necessarily internally). But thanks a lot for highlighting `context` <-> mutablespan link, I will look further.

> So, I think we need to focus on what is the pattern suggested for migrating to ReentrantLock from synchronized, when in synchronized we can use the object itself as the lock. Be careful to not expose API surface when thinking this through!

`ReentrantLock` sadly does not sadly help us much here since `synchronized(state)` in brute force would be equivalent to something like `state.lock.lock(); ....` - exactly the problem you mentioned, API surface will be harmed.

Reply:
Why don't we solve for the issue then and leave the rest alone?

So, the only reason why state is held for a long time (microseconds) is when it is being sized for encoding in the async reporter, which is in another project, or someone is doing large amount of work tagging (bug).

One option could be that when a configuration is present, use copy mode which will be faster than microseconds (faster than sizing the span as you don't need to check encoded size of strings etc).

Also, the reported stack trace is inherently inefficient as it is copying the span from brave's type to zipkin's type first (`ZipkinSpanHandler`). There's a brave-native encoder which is already queued for micrometer 1.3 (`https://github.com/micrometer-metrics/tracing/releases/tag/v1.3.0-M2

```
    java.base/java.util.concurrent.locks.ReentrantLock$Sync.lock(Unknown Source)
    java.base/java.util.concurrent.locks.ReentrantLock.lock(Unknown Source)
    zipkin2.reporter.ByteBoundedQueue.offer(ByteBoundedQueue.java:51)
    zipkin2.reporter.AsyncReporter$BoundedAsyncReporter.report(AsyncReporter.java:259)
    zipkin2.reporter.brave.ConvertingSpanReporter.report(ConvertingSpanReporter.java:44)
    zipkin2.reporter.brave.ConvertingSpanReporter.report(ConvertingSpanReporter.java:29)
    zipkin2.reporter.brave.ZipkinSpanHandler.end(ZipkinSpanHandler.java:148)
    brave.internal.handler.NoopAwareSpanHandler$CompositeSpanHandler.end(NoopAwareSpanHandler.java:107)
    brave.internal.handler.NoopAwareSpanHandler.end(NoopAwareSpanHandler.java:57)
```

Finally, I really think we owe it to define what is expensive blocking and so we can measure it. We are assuming this is expensive just because some message ended up in the log of a suboptimal setup.

Reply:
basically the spring community have spent a lot of time on things but not making things like this more efficient. If someone wanted to, they could make an alternative span handler that doesn't size first, or copies as it is. Instead, we are in a situation where it is assumed you and I have to do all and any work due to pressure caused by external factors. Great to be the slave of OSS!

Reply:
tracing in micrometer is very slow due to all the frameworks used to the point where overhead vs other options is milliseconds worse, not microseconds. Staring at this issue as if it holistically changes things is highly suspicious and a questionable reason to introduce a lot of risk. I think the most narrow solution to the problem, including muting log messages is a decent thing to consider.

e.g. measure this and you'll see hello world requests will millis extra overhead. https://github.com/openzipkin/brave-example/tree/master/webflux6-micrometer